### PR TITLE
Extract HackyAI logic into *Module traits - part 1

### PR DESCRIPTION
**Pull request by reaperrr**
This allows to disable or replace them at will.

This first PR turns the already existing harvester manager and support power manager into player traits, and directly extracts the capture logic and reaction-to-damage logic to traits as well.

The yaml changes were originally 100% update rule output, but I removed redundancies later (most AIs in each mod used the exact same SupportPower settings).

The only module I haven't really tested yet is the capture module, due to lack of testcases in our shipping mods.

Other than that, everything including the update rule has been tested and should work as intended.

### CODE REVIEW

File: OpenRA.Mods.Common/UpdateRules/UpdatePath.cs [hunk 1]
```diff
@@ -86,7 +86,7 @@ public class UpdatePath
 
 			new UpdatePath("playtest-20180729", new UpdateRule[]
 			{
-				// Bleed only changes here
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/UpdateRules/UpdatePath.cs]: Think we want to keep this comment.

Comment (reaperrr) [reply OpenRA.Mods.Common/UpdateRules/UpdatePath.cs]: Oops, yes.

File: OpenRA.Mods.Common/AI/HackyAI.cs [hunk 2]
```diff
@@ -270,14 +236,17 @@ public CPos GetRandomBaseCenter()
...
+		CPos defenseCenter;
 		PowerManager playerPower;
 		PlayerResources playerResource;
 		int ticks;
 
 		BitArray resourceTypeIndices;
 
-		AIHarvesterManager harvManager;
-		AISupportPowerManager supportPowerManager;
+		AIHarvesterModule harvModule;
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/HackyAI.cs]: Extracting these to modules but then keeping the module types hardcoded hides a lot of the potential benefits here.  Would it be feasible to have a single generic `IAIModule` interface and then define a `Dictionary<string, string>` in the `HackyAIInfo` for condition: enable module name?  Or are we still relying too much on calling into the internal details of the classes?  If we are, then please lets use interfaces for each type.

Comment (reaperrr) [reply OpenRA.Mods.Common/AI/HackyAI.cs]: > Or are we still relying too much on calling into the internal details of the classes?

I'm afraid so, but I'll check if I can reduce that further.

**Comment by reaperrr** [comment]
> Is it just me, or the SupportPower modules in TD/RA are defined wrong - you seem to aim for sharing them between the AIs but then each AI's using their own name to assign them.

Look at the respective "Remove..." commits, they fix that.
The first yaml commit is just the unmodified update rule output. It would be way too complicated to check for duplication during the update rule, so I just extract a support power module per AI.
The 2nd commit for each mod then removes the redundancy manually.

**Comment by GraionDilach** [comment]
Ahyeah, I can see now - D2k is the bugged one.

**Comment by Gerseras** [comment]
reaperr are you still working on this? If so are there ways to help you as i would like to have the ai modular so we can implement a own/custom basebuilder module for the kknd mods so the ai learns to handle the other way of constructing buildings

**Comment by reaperrr** [comment]
> reaperr are you still working on this? If so are there ways to help you as i would like to have the ai modular so we can implement a own/custom basebuilder module for the kknd mods so the ai learns to handle the other way of constructing buildings

I've been working mostly on other things in the meantime, but I think I can whip this into good-enough shape relatively quickly.

**Comment by reaperrr** [comment]
Rebased and updated.

@Gerseras @IceReaper Feel free to help reviewing/testing now :)

Locally I have the BaseManagerModule ready as well (except for the update rule and yaml changes, but I guess this probably won't get merged immediately anyway).

**Comment by Gerseras** [comment]
@reaperrr We will take a look at it next week and try to use a custom HarvesterModule for our Tanker logic
if that works we would love to have the basebuilder module as that could become quite intresting

**Comment by pchote** [comment]
We have been a bit inconsistent with "AI" vs "Bot" before now - we have tended to use "Bot" in the code, and "AI" for user facing things, except for `HackyAI`.  IMO it would be better to call these `IBotModule`s as this is a bit more specific than "AI", which could also refer to unit-level AI on player-owned units.

File: OpenRA.Mods.Common/AI/HackyAI.cs [hunk 3]
```diff
@@ -162,8 +162,8 @@ public class BuildingCategories
 		[Desc("Terrain types which are considered water for base building purposes.")]
 		public readonly HashSet<string> WaterTerrainTypes = new HashSet<string> { "Water" };
 
-		[Desc("Avoid enemy actors nearby when searching for a new resource patch. Should be somewhere near the max weapon range.")]
-		public readonly WDist HarvesterEnemyAvoidanceRadius = WDist.FromCells(8);
+		[Desc("Which harvester handling module to use. Leave empty to disable.")]
+		public readonly string HarvesterModule = "default-harvester-module";
```
Comment (pchote) [inline OpenRA.Mods.Common/AI/HackyAI.cs]: Having specifically behaving modules referred to via strings like this sets off warning bells in my head.  What if someone references a support power module type here, or something different?

If we want to refer to specific types of module here, then we should come up with a way to annotate them - perhaps using a new attribute with a tag type.

A much better fix would be to not have specific `*Modules` at all, and instead a generic `Modules` list, with `HackyAI` itself not caring about what each module is/does.

File: OpenRA.Mods.Common/AI/AIDamageReactionModule.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+using OpenRA.Support;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.AI
+{
+	[Desc("Manages how AI reacts to its actors receiving damage.")]
+	public class AIDamageReactionModuleInfo : IAIModuleInfo
+	{
+		[Desc("Name for identification purposes.")]
+		public readonly string Name = "default-damagereaction-module";
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/AI/AIDamageReactionModule.cs]: Why is `default` in this name?

Comment (Phrohdoh) [reply OpenRA.Mods.Common/AI/AIDamageReactionModule.cs]: If we're going to have modules added by default shouldn't the harvester one also be a default?

File: OpenRA.Mods.Common/AI/AIHarvesterModule.cs [hunk 5]
```diff
@@ -20,54 +20,52 @@
 
 namespace OpenRA.Mods.Common.AI
 {
-	class AIHarvesterManager
+	[Desc("Manages AI harvesters to ensure they always continue harvesting as long as there's resources on the map.")]
+	public class AIHarvesterModuleInfo : IAIModuleInfo
 	{
-		readonly HackyAI ai;
+		[Desc("Name for identification purposes.")]
+		public readonly string Name = "default-harvester-module";
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/AI/AIHarvesterModule.cs]: IIRC the other PR doesn't have `default-` in the harvester module's name, which do we want?

File: mods/d2k/rules/ai.yaml [hunk 6]
```diff
@@ -332,6 +250,9 @@ Player:
 			carryall: 4
 		SquadSize: 10
 		MaxBaseRadius: 40
+	AIHarvesterModule:
+	AIDamageReactionModule:
+	AISupportPowerModule:
 		SupportPowerDecisions:
```
Comment (Phrohdoh) [inline mods/d2k/rules/ai.yaml]: Nit: Could we rename this property? No need, IMO, to have `SupportPower` duplicated now that this lives under a support-power-specific trait.

File: OpenRA.Mods.Common/AI/AIDamageReactionModule.cs (hunk 4 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/AI/AIDamageReactionModule.cs]: That was a relic from earlier code revisions.

File: OpenRA.Mods.Common/AI/AIHarvesterModule.cs (hunk 5 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/AI/AIHarvesterModule.cs]: The plumbing has out-evolved this Name property in the split off PRs

File: mods/d2k/rules/ai.yaml (hunk 6 above)
Comment (reaperrr) [reply mods/d2k/rules/ai.yaml]: Will do that in #15786.

**Comment by reaperrr** [comment]
This is both too monolithic and too stale by now, I'll split off the modules in separate PRs. Closing.

**Pull request closed without merging**